File: src-picker/dialog.cpp
/**
 * Approval dialog implementation.
 *
 * Shows a Qt6 dialog for screen recording permission, with fallback
 * to hyprland-dialog if needed.
 */

#include "dialog.h"

#include <QApplication>
#include <QDialog>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QIcon>
#include <QStyle>
#include <QFont>
#include <QSizePolicy>
#include <QScreen>
#include <QProcess>
#include <QRandomGenerator>
#include <iostream>

/**
 * Internal result enum for the dialog.
 */
enum class InternalResult {
    AlwaysAllow,
    AllowOnce,
    Deny
};

/**
 * Permission dialog widget.
 */
class PermissionDialog : public QDialog {
public:
    PermissionDialog(const QString& sourceDesc, QWidget* parent = nullptr) 
        : QDialog(parent), result_(InternalResult::Deny) 
    {
        setWindowTitle("OmniRec - Screen Recording Permission");
        setMinimumWidth(400);
        
        // Set window flags for floating behavior in tiling WMs:
        // Qt::Dialog is the standard dialog type that compositors float by default
        // Qt::WindowStaysOnTopHint keeps it above other windows
        setWindowFlags(Qt::Dialog | Qt::WindowStaysOnTopHint);
        
        // Set window modality to make it behave as a modal dialog
        setWindowModality(Qt::ApplicationModal);
        
        // Fixed size policy prevents tiling WMs from resizing
        setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
        
        auto* layout = new QVBoxLayout(this);
        layout->setSpacing(16);
        layout->setContentsMargins(24, 24, 24, 24);
        
        // Icon and title row
        auto* headerLayout = new QHBoxLayout();
        
        auto* iconLabel = new QLabel(this);
        QIcon warningIcon = style()->standardIcon(QStyle::SP_MessageBoxQuestion);
        iconLabel->setPixmap(warningIcon.pixmap(48, 48));
        headerLayout->addWidget(iconLabel);
        
        auto* titleLabel = new QLabel("Allow OmniRec to record your screen?", this);
        QFont titleFont = titleLabel->font();
        titleFont.setPointSize(titleFont.pointSize() + 2);
        titleFont.setBold(true);
        titleLabel->setFont(titleFont);
        titleLabel->setWordWrap(true);
        headerLayout->addWidget(titleLabel, 1);
        
        layout->addLayout(headerLayout);
        
        // Source description
        auto* descLabel = new QLabel(sourceDesc, this);
        descLabel->setWordWrap(true);
        descLabel->setStyleSheet("color: #666; padding-left: 64px;");
        layout->addWidget(descLabel);
        
        // Spacer
        layout->addSpacing(8);
        
        // Buttons
        auto* buttonLayout = new QHBoxLayout();
        buttonLayout->setSpacing(8);
        
        auto* denyBtn = new QPushButton("Deny", this);
        denyBtn->setMinimumWidth(100);
        connect(denyBtn, &QPushButton::clicked, this, [this]() {
            result_ = InternalResult::Deny;
            reject();
        });
        
        auto* allowOnceBtn = new QPushButton("Allow Once", this);
        allowOnceBtn->setMinimumWidth(100);
        connect(allowOnceBtn, &QPushButton::clicked, this, [this]() {
            result_ = InternalResult::AllowOnce;
            accept();
        });
        
        auto* alwaysAllowBtn = new QPushButton("Always Allow", this);
        alwaysAllowBtn->setMinimumWidth(120);
        alwaysAllowBtn->setDefault(true);
        alwaysAllowBtn->setStyleSheet(
            "QPushButton { background-color: #2196F3; color: white; font-weight: bold; }"
            "QPushButton:hover { background-color: #1976D2; }"
        );
        connect(alwaysAllowBtn, &QPushButton::clicked, this, [this]() {
            result_ = InternalResult::AlwaysAllow;
            accept();
        });
        
        buttonLayout->addStretch();
        buttonLayout->addWidget(denyBtn);
        buttonLayout->addWidget(allowOnceBtn);
        buttonLayout->addWidget(alwaysAllowBtn);
        
        layout->addLayout(buttonLayout);
    }
    
    InternalResult result() const { return result_; }
    
private:
    InternalResult result_;
};

/**
 * Format source description for display.
 */
static QString formatSourceDesc(const QString& sourceType, const QString& sourceId)
{
    if (sourceType == "monitor") {
        return QString("Display: %1").arg(sourceId);
    } else if (sourceType == "window") {
        return QString("Window: %1").arg(sourceId);
    } else if (sourceType == "region") {
        return QString("Region on: %1").arg(sourceId);
    } else {
        return QString("Source: %1").arg(sourceId);
    }
}

/**
 * Try to show dialog using hyprland-dialog (fallback).
 */
static DialogResult tryHyprlandDialog(const QString& sourceDesc)
{
    std::cerr << "[omnirec-picker] Trying hyprland-dialog (fallback)" << std::endl;
    
    QString text = QString("OmniRec is requesting permission to record your screen.\n\n%1").arg(sourceDesc);
    
    QProcess process;
    process.start("hyprland-dialog", QStringList() 
        << "--title" << "OmniRec - Screen Recording Permission"
        << "--text" << text
        << "--buttons" << "Always Allow;Allow Once;Deny");
    
    if (!process.waitForStarted(5000)) {
        std::cerr << "[omnirec-picker] hyprland-dialog not found" << std::endl;
        return DialogResult::Denied;
    }
    
    process.waitForFinished(-1);
    
    QString response = QString::fromUtf8(process.readAllStandardOutput()).trimmed();
    std::cerr << "[omnirec-picker] hyprland-dialog response: '" << response.toStdString() 
              << "' (exit: " << process.exitCode() << ")" << std::endl;
    
    if (response == "Always Allow") {
        std::cerr << "[omnirec-picker] User approved with always_allow=true" << std::endl;
        return DialogResult::AlwaysAllow;
    } else if (response == "Allow Once") {
        std::cerr << "[omnirec-picker] User approved with always_allow=false" << std::endl;
        return DialogResult::AllowOnce;
    } else {
        std::cerr << "[omnirec-picker] User denied" << std::endl;
        return DialogResult::Denied;
    }
}

DialogResult showApprovalDialog(const QString& sourceType, const QString& sourceId)
{
    std::cerr << "[omnirec-picker] show_approval_dialog called" << std::endl;
    
    QString sourceDesc = formatSourceDesc(sourceType, sourceId);
    
    // Check if we have a QApplication (we should, since main creates one)
    if (!QApplication::instance()) {
        std::cerr << "[omnirec-picker] No QApplication, falling back to hyprland-dialog" << std::endl;
        return tryHyprlandDialog(sourceDesc);
    }
    
    // Show our Qt dialog
    PermissionDialog dialog(sourceDesc);
    dialog.setObjectName("omnirec-dialog");
    
    // Center on screen and lock size to prevent tiling
    dialog.adjustSize();
    dialog.setFixedSize(dialog.size());  // Lock size after layout calculation
    
    if (QScreen* screen = QApplication::primaryScreen()) {
        QRect screenGeometry = screen->availableGeometry();
        int x = (screenGeometry.width() - dialog.width()) / 2 + screenGeometry.x();
        int y = (screenGeometry.height() - dialog.height()) / 2 + screenGeometry.y();
        dialog.move(x, y);
    }
    
    dialog.exec();
    
    switch (dialog.result()) {
        case InternalResult::AlwaysAllow:
            std::cerr << "[omnirec-picker] User approved with always_allow=true" << std::endl;
            return DialogResult::AlwaysAllow;
        case InternalResult::AllowOnce:
            std::cerr << "[omnirec-picker] User approved with always_allow=false" << std::endl;
            return DialogResult::AllowOnce;
        case InternalResult::Deny:
        default:
            std::cerr << "[omnirec-picker] User denied" << std::endl;
            return DialogResult::Denied;
    }
}

QString generateApprovalToken()
{
    // Generate 32 random bytes (256 bits)
    QByteArray bytes(32, 0);
    QRandomGenerator* rng = QRandomGenerator::global();
    for (int i = 0; i < 32; ++i) {
        bytes[i] = static_cast<char>(rng->bounded(256));
    }
    
    // Convert to hex string
    return bytes.toHex();
}

File: src-picker/src/dialog.rs
//! Approval dialog for screen recording consent.
//!
//! This module provides a Qt6 dialog that asks the user to approve
//! OmniRec's screen recording request.

use std::path::PathBuf;
use std::process::{Command, Stdio};

/// Result of the approval dialog.
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum DialogResult {
    /// User approved the recording request.
    Approved {
        /// Whether "Always allow" was checked.
        always_allow: bool,
    },
    /// User denied the recording request.
    Denied,
}

/// Get the path to the omnirec-dialog binary.
/// 
/// Searches in order:
/// 1. Next to the picker binary (for development)
/// 2. In the same directory as the picker (installed)
/// 3. In PATH
fn find_dialog_binary() -> Option<PathBuf> {
    // Check next to this binary (dev build location)
    if let Ok(exe) = std::env::current_exe() {
        // Development: qt-dialog/build/omnirec-dialog relative to picker binary
        if let Some(picker_dir) = exe.parent() {
            // picker is at src-picker/target/release/omnirec-picker
            // dialog is at src-picker/qt-dialog/build/omnirec-dialog
            let dev_path = picker_dir
                .parent() // target
                .and_then(|p| p.parent()) // src-picker
                .map(|p| p.join("qt-dialog").join("build").join("omnirec-dialog"));
            
            if let Some(path) = dev_path {
                if path.exists() {
                    return Some(path);
                }
            }
            
            // Installed: same directory as picker
            let installed_path = picker_dir.join("omnirec-dialog");
            if installed_path.exists() {
                return Some(installed_path);
            }
        }
    }
    
    // Check PATH
    if let Ok(output) = Command::new("which").arg("omnirec-dialog").output() {
        if output.status.success() {
            let path = String::from_utf8_lossy(&output.stdout).trim().to_string();
            if !path.is_empty() {
                return Some(PathBuf::from(path));
            }
        }
    }
    
    None
}

/// Show the approval dialog and wait for user response.
///
/// Uses our custom Qt6 dialog (omnirec-dialog) for a polished UI.
/// Falls back to hyprland-dialog if omnirec-dialog is not available.
///
/// # Arguments
/// * `source_type` - Type of capture source (e.g., "monitor", "window", "region")
/// * `source_id` - Identifier of the capture source (e.g., "DP-1", window title)
pub fn show_approval_dialog(source_type: &str, source_id: &str) -> DialogResult {
    eprintln!("[omnirec-picker] show_approval_dialog called");

    let source_desc = match source_type {
        "monitor" => format!("Display: {}", source_id),
        "window" => format!("Window: {}", source_id),
        "region" => format!("Region on: {}", source_id),
        _ => format!("Source: {}", source_id),
    };

    // Try our custom Qt dialog first
    if let Some(result) = try_omnirec_dialog(&source_desc) {
        return result;
    }

    // Fall back to hyprland-dialog
    if let Some(result) = try_hyprland_dialog(&source_desc) {
        return result;
    }

    // No dialog tool available - deny by default for security
    eprintln!("[omnirec-picker] ERROR: No dialog tool available!");
    eprintln!("[omnirec-picker] Please ensure omnirec-dialog is built or hyprland-guiutils is installed");
    DialogResult::Denied
}

/// Try to show dialog using our custom omnirec-dialog Qt binary.
fn try_omnirec_dialog(source_desc: &str) -> Option<DialogResult> {
    let dialog_path = find_dialog_binary()?;
    
    eprintln!("[omnirec-picker] Using omnirec-dialog at {:?}", dialog_path);

    let result = Command::new(&dialog_path)
        .arg(source_desc)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output();

    match result {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let response = stdout.trim();
            eprintln!(
                "[omnirec-picker] omnirec-dialog response: '{}' (exit: {:?})",
                response,
                output.status.code()
            );

            match response {
                "ALWAYS_ALLOW" => {
                    eprintln!("[omnirec-picker] User approved with always_allow=true");
                    Some(DialogResult::Approved { always_allow: true })
                }
                "ALLOW_ONCE" => {
                    eprintln!("[omnirec-picker] User approved with always_allow=false");
                    Some(DialogResult::Approved { always_allow: false })
                }
                _ => {
                    eprintln!("[omnirec-picker] User denied");
                    Some(DialogResult::Denied)
                }
            }
        }
        Err(e) => {
            eprintln!("[omnirec-picker] omnirec-dialog failed: {}", e);
            None
        }
    }
}

/// Try to show dialog using hyprland-dialog (fallback).
fn try_hyprland_dialog(source_desc: &str) -> Option<DialogResult> {
    eprintln!("[omnirec-picker] Trying hyprland-dialog (fallback)");

    let text = format!(
        "OmniRec is requesting permission to record your screen.\n\n{}",
        source_desc
    );

    let result = Command::new("hyprland-dialog")
        .args([
            "--title",
            "OmniRec - Screen Recording Permission",
            "--text",
            &text,
            "--buttons",
            "Always Allow;Allow Once;Deny",
        ])
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output();

    match result {
        Ok(output) => {
            let stdout = String::from_utf8_lossy(&output.stdout);
            let response = stdout.trim();
            eprintln!(
                "[omnirec-picker] hyprland-dialog response: '{}' (exit: {:?})",
                response,
                output.status.code()
            );

            match response {
                "Always Allow" => {
                    eprintln!("[omnirec-picker] User approved with always_allow=true");
                    Some(DialogResult::Approved { always_allow: true })
                }
                "Allow Once" => {
                    eprintln!("[omnirec-picker] User approved with always_allow=false");
                    Some(DialogResult::Approved { always_allow: false })
                }
                _ => {
                    eprintln!("[omnirec-picker] User denied");
                    Some(DialogResult::Denied)
                }
            }
        }
        Err(e) => {
            if e.kind() == std::io::ErrorKind::NotFound {
                eprintln!("[omnirec-picker] hyprland-dialog not found");
                None
            } else {
                eprintln!("[omnirec-picker] hyprland-dialog failed: {}", e);
                Some(DialogResult::Denied)
            }
        }
    }
}

/// Generate a random 256-bit approval token as a hex string.
pub fn generate_approval_token() -> String {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    let bytes: [u8; 32] = rng.gen();
    bytes.iter().map(|b| format!("{:02x}", b)).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_token_length() {
        let token = generate_approval_token();
        assert_eq!(token.len(), 64); // 32 bytes * 2 hex chars
    }

    #[test]
    fn test_generate_token_uniqueness() {
        let token1 = generate_approval_token();
        let token2 = generate_approval_token();
        assert_ne!(token1, token2);
    }

    #[test]
    fn test_generate_token_is_hex() {
        let token = generate_approval_token();
        assert!(token.chars().all(|c| c.is_ascii_hexdigit()));
    }
}
